docs/adr/0006-server-arch.md

```diff
@@ -0,0 +1,185 @@
+        |     +------------+       +-----------+
+        |     | aggregates |       | services  |
+        |     +------------+       +-----------+
+        |           |(n)
+    +------------------------------------------+
+    |                 models                   |
+    +------------------------------------------+
+    |             data access layer            |
+    +------------------------------------------+
+
+</pre>
+
+---
+
+## General Principles
+
+Each of the following sections describes the general principles that are to
+be used when writing code. This approach uses SOLID Principles.
+
+### GraphQL Resolvers
+
+- The external interface uses graphQL and these are handled with the resolvers.
+- The resolvers should contain no logic themselves. Rather this is delegated
+  to a `use-case` making the resolvers a thin wrapper around the controller logic.
+- Rationale is that it should be easy to implement another interface on the same
+  logic.
+
+#### Example
+
+```
+const { getSubmissionUseCase, } = require('use-cases')
+
+const resolvers = {
+  Query: {
+    async manuscript(_, { id }, { user }) {
+      return getSubmissionUseCase(id, user)
+    },
+    ...
+  }
+  ...
+```
+
+### Use-Cases
+
+- In general this layer will have an interface that maps one-to-one with the
+  graphQL layer.
+- For a given use-case it will require certain services and config - these
+  dependencies will be created by this layer and injected (when needed) into
+  the aggregates / models.
+
+#### Example
+
+```js
+const config = require('config')
+const Storage = require('MyStorageService')
+const Submission = require('aggregates/Submission')
+
+const User = require('@elifesciences/component-model').UserModel
+const File = require('@elifesciences/component-model').FileModel
+const Team = require('@elifesciences/component-model').TeamModel
+const Manuscript = require('@elifesciences/component-model').ManuscriptModel
+
+async getSubmissionUseCase(manuscriptId, user) => {
+  checkPermissions(id, user)
+  const models = { ... }
+  const services = { Storage, ... }
+  const submission = new Submision(config, models, services)
+  return submission.get(manuscriptId, user).toJSON()
+}
+
+module.exports = {
+  getSubmissionUseCase,
+}
+```
+
+### The Aggregate
+
+- [See DDD_Aggregates](https://www.martinfowler.com/bliki/DDD_Aggregate.html)
+- An aggregate
+
+```js
+// Note all dependencies injected.
+
+class Submission {
+  constructor(config, models, services) {
+    this.config = config
+    this.models = models
+    this.storage = services.Storage
+  }
+
+  async get(id, user) {
+      /*
+      * Here we could get the files' contents, the teams and the manuscript
+      * then stitch them together to send back
+      */
+      return {
+        manuscript: await this.models.Manuscript.find(id),
+        teams: await this.models.Team.findByManuscriptId(id),
+        files: await this.models.File.get(config, fileId, this.services.Storage)
```

Comment:
some of these are conceptually different, may be worth noting. `manuscript` and `files` are internal to this Aggregate whereas `teams` seems to be a reference to other Aggregates. Therefore `teams` should be conceptually read-only in this Aggregate's view.

Reply:
This is good as it reflects a deeper-level of thinking than I had given this - thanks.

Reply:
The next commit will deepen this thinking with considering `SuggestedEditors`

Reply:
@giorgiosironi So are you saying that a Team should be its own aggregate, enforce its own consistency checks and modifying its state should be done via its methods ?

Reply:
`Team` is not a domain model ... its a data model and should not "escape" the Data Layer ... its an implementation of storing people associated with a manuscript.